### PR TITLE
Support Primitive List feature

### LINKED ISSUE #575: RealmList<Integer / String / ...>
Dear realm-Team, 

I'm really impressed by Realm and I do like the way you're providing support and solutions on github, SO and gGroups! I like working with Realm - however there are three really big drawbacks for me:

1) No model-instantiation by `= new ...` - I know you're on it. I can't wait for this! Do we have to wait for the "whole" JSON-Upgrade or will there be an earlier release providing this feature?

2) No model-passing across threads. That's hindering all the great benefits of EventBus and Android Priority Job Queue. But I read that it will be possible in the further versions and I'm looking forward on this.

3) **The reason for this issue:**
In my model class I'm allowed to have primitive fields as well as object fields of objects that extend `RealmObject`.
But I'm not allowed to have a field type like `RealmList<Integer>` or `RealmList<String>`.
However it is possible to work around this by

``` java
RealmList<RealmString> names;
```

and

``` java
public class RealmString extends RealmObject {
private String name;
// getter & setter
```

It works but it's _really_ ugly and should be unnecessary in my opinion.
Is there any improvement planned regarding this?

Thanks again for your great work - I'm glad Realm helps me developing my app!

--- 

Related PRs:
 - [ ] API design (#4165)
 - [ ] Implementation (#5031)

### PR DESCRIPTION
**Pull request by zaki50**
This PR implements #4165 and fixes #575.

TODO: 
**Required:**
- [x] Add, modify APIs (@zaki50) #5145 
- [x] Loosen type restriction on `RealmList<RealmModel>` to `RealmList<Object>` (@zaki50) #5145
- [x] Loosen type restriction on `RealmResults<RealmModel>` to `RealmResults<Object>` (@zaki50) #5145
- [x] Implement type parameter check in annotation processor (@zaki50) #5168 
- [x] Implement Schema related feature used on initialization (@zaki50) #5209 (part 7)
- [x] Implement proxy's getters/setters logic in annotation processor (@zaki50) #5210 (part 8)
- [x] Implement support for primitives in `RealmList` (Java side, @zaki50) #5213 (part 9)
- [x] Wire with Core/Object Store (@beeender) #5208 (part 6), #5216 (part10)
- [x] Fix for change listener (@beeender) #5259 (part 13) 
- [x] Fix a bug introduced in this feature branch (@zaki50) #5266 
- [x] If we merge feature branch without optional feature, we need to add a code to throw an exception to some methods which does not support primitive list (ex. `createSnapshot()`) (@zaki50) #5336 (part 15)

**Optional/In prioritized order:**
- [x] `copyToRealm/copyOrUpdate` (@zaki50) #5210 (part 8)
- [x] `insert/insertOrUpdate` (@zaki50) #5210 (part 8)
- [x] `copyFromRealm` (@zaki50) #5210 (part 8)
- [ ] APIs for migration (mainly methods in `RealmObjectSchema`)  (@cmelchior) #5329 
- [ ] `OrderedRealmCollectionImpl`, `OrderedRealmCollectionSnapshot`  (@zaki50) #5220 (part11)
- [ ] Dynamic API (@zaki50) #5221 (part 12)
- [ ] Query APIs for list elements
- [ ] Query APIs for object (by using primitive list value)
- [ ] Implement support for primitives in `RealmResults` (@zaki50) #5220 (part11)
- [ ] JSON APIs

### CODE REVIEW

**Comment by zaki50** [comment]
I've merged Part 1 to Part 5  into this PR.

File: realm/realm-library/src/androidTest/java/io/realm/RealmSchemaTests.java [hunk 1]
```diff
@@ -606,4 +606,48 @@ public void schemaInformationOfPrimitiveLists() {
             assertFalse(objectSchema.isPrimaryKey(fieldName));
         }
     }
+
+	@Test
```
Comment (Zhuinden) [inline realm/realm-library/src/androidTest/java/io/realm/RealmSchemaTests.java]: indentation

Comment (zaki50) [reply realm/realm-library/src/androidTest/java/io/realm/RealmSchemaTests.java]: Thanks. Fixed

**Comment by VenomVendor** [comment]
Great relief. Updating my code base to add Primitive support.

**Comment by cmelchior** [comment]
@VenomVendor Keep in mind that queries are not supported in the initial release, but we will add them as soon as possible

**Comment by Zhuinden** [comment]
@cmelchior and same for migration support for adding/removing RealmList of primitives to existing schema

**Comment by jonathansds** [comment]
How to use? It doesn't compile if I just use RealmList<Long>.

**Comment by Zhuinden** [comment]
@jonathansds `RealmList<String>` and stuff like that.

**Comment by jonathansds** [comment]
It doesn't compile when I use RealmList < Long >

**Comment by jonathansds** [comment]
It says that "Long" doesn't extends RealmObject

**Comment by Zhuinden** [comment]
@jonathansds are you using `4.0.0-RC1`?

**Comment by jonathansds** [comment]
That was it! Sorry, missed that.

**Comment by hardysim** [comment]
So this is very nice but it's limited to `java.lang.String` - can we get `kotlin.String` as well (and the other types)?

**Pull request merged**